### PR TITLE
[no-ci] [precompile] Serve the continuations Dynamo never traced as plain Python

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #198217
* #198227
* #198138
* #198216
* #198137
* #198215
* #198136
* __->__ #198214
* #198243
* #198135
* #198213
* #197993
* #198223
* #198211

**Problem.** When a captured step ends in `.backward()`, the continuation Dynamo resumes after that call reaches no tensor, so `convert_frame` skips it before compiling and it runs as plain Python during capture. Its record still exists (the frame ahead names it by resume name) but has no variant, and a standalone artifact treated that as a coverage gap: `_serving_mode` sent the capture to installed serving and the driver bound a refusal under the resume name.

**Fix.** The record now says `trivial` for a code entry Dynamo never traced, `_serving_mode` counts such a non-entry frame as covered, and the driver rebuilds it as a plain function over its code object in the captured module's scope, so the frame runs at serve time exactly as it ran during capture.

**Summary.** A standalone multi-graph artifact rebuilds each captured frame from its record, and a continuation with no variant used to be read as a frame the artifact could not serve. That was wrong for the continuation after a trailing `.backward()`: Dynamo skipped it before tracing, it ran eager during capture, and yet the capture was sent to installed serving while a call reaching the frame in a standalone rebuild would have raised. `_multigraph_frames` now records `trivial = not code.has_compile_id and not code.bypassed`, `_serving_mode` treats a trivial non-entry frame as covered, and `_build_multigraph_forward` binds such a frame as a `types.FunctionType` over its code object in the captured module's scope (a factory over the closure tuple when it has free variables, the way Dynamo binds one). The design constraint is that only a frame Dynamo never entered qualifies -- the entry never does, and a bypassed frame never does -- because those are the frames whose eager execution is what capture actually exercised. A continuation Dynamo compiled but kept no variant of is unchanged: it still sends the capture to installing and still refuses at the call that reaches it, since that is a coverage gap rather than plain Python.

## The frame Dynamo never traced

The new test captures a step whose last statement is `.backward()`:

```python
def step(model, x):
    (model(x) * _MULTIGRAPH_SCALE).sum().backward()

package = CompilePackage(step)
compiled = torch._dynamo.optimize(backend="eager", package=package, guard_filter_fn=default_guard_filter_fn)(step)
compiled(model, x)
frames = _multigraph_frames(package.cache_entry())
```

`.backward()` is a graph break, so Dynamo mints a resume function for the rest of `step`. That continuation only returns `None`; no tensor reaches it, so `convert_frame` skips it before compiling and it runs as plain Python during capture. Its `CompilePackage` code entry never went through `code_context`, so `has_compile_id` stays `False`, and it is not bypassed. The records now read:

```python
[f["trivial"] for f in frames]        # [False, True]
[len(f["variants"]) for f in frames]  # [1, 0]
```

The parent produced the same two records minus the `trivial` key, and the second one -- reachable from the entry's bytecode, no variant -- was indistinguishable from a continuation Dynamo compiled but could not keep.

## Recording `trivial`

```python
# torch/_precompile.py, _multigraph_frames
{
    "is_entry": i == 0,
    "bypassed": code.bypassed,
    # now: never entered Dynamo (skipped before tracing), so it ran eager
    # during capture and the driver runs it eager too
    "trivial": not code.has_compile_id and not code.bypassed,
    "code": code.python_code,
    ...
}
```

`has_compile_id` is set by `CompilePackage.code_context` when Dynamo actually traces a code object, so a frame that was skipped keeps it `False`. A bypassed frame is excluded on purpose: its guarded codes are dead for a different reason (its guards could not be serialized, or a backend artifact was missing), and install keeps its binding but none of its codes.

## Counting it as covered

```python
# torch/_precompile.py, _serving_mode
reachable = _reachable_frames(frames)
# parent: covered = {i for i, frame in enumerate(frames) if frame["variants"]}
covered = {
    i
    for i, frame in enumerate(frames)
    if frame["variants"] or (frame.get("trivial") and not frame["is_entry"])
}
return "standalone" if covered == reachable else "installed"
```

The entry is excluded because a trivial entry means the whole capture ran eager, which is not something an artifact should serve. The test pins both verdicts on the same records:

```python
_serving_mode(frames)                  # now: "standalone" (parent: "installed")
frames[1]["trivial"] = False
_serving_mode(frames)                  # "installed": a compiled continuation with no variant is still a gap
```

## Rebuilding it as plain Python

```python
# torch/_precompile_driver.py, _make_dispatcher
if not frame["variants"] and frame.get("trivial") and not is_entry:
    scope = _scope(frame)

    def _plain(closure):
        return types.FunctionType(target, scope, target.co_name, None, closure)

    if target.co_freevars:
        return _plain
    return _plain(None)
if not frame["variants"]:
    ... # parent: this branch bound a refusal under the resume name
```

The branch sits ahead of the existing no-variant refusal, so a trivial continuation never reaches it. `_scope(frame)` opens the captured module so the bytecode's `LOAD_GLOBAL`s read the module's live dict, the same scope a compiled variant is bound over. When the code object closes over locals, Dynamo binds the continuation as a factory taking the closure tuple and the frame ahead passes one, so the driver returns `_plain` itself in that case and a bound function otherwise -- mirroring the `factory` / `dispatch` split the compiled path already uses.

The refusal branch that remains is still the right answer for the other no-variant continuation: one Dynamo compiled (`has_compile_id` set) but kept no guarded code of. In a standalone artifact such a record is dead unless it is the entry, and a dead continuation must still bind its resume names, so its refusal is deferred to a call.

## Tests

`test_trivial_continuation_is_served_as_plain_python` captures the `step` above, checks the `[False, True]` / `[1, 0]` records, checks the `_serving_mode` verdict before and after flipping `trivial` on the continuation, then rebuilds the driver from the records in this process -- `exec` of `_build_multigraph_forward`'s source over the pickled frames and backends, with the captured module aliased into `sys.modules` -- and serves the step on a fresh `Linear` loaded from the captured model's state dict, comparing `weight.grad` and `bias.grad` with an eager run of `step`.

Test Plan:

`test_trivial_continuation_is_served_as_plain_python` captures a step whose last statement is `.backward()`, checks the record and both `_serving_mode` verdicts, rebuilds the driver from the records and serves the step in this process, comparing the runtime model's grads with eager.

```
python test/test_precompile.py
lintrunner -a torch/_precompile.py torch/_precompile_driver.py test/test_precompile.py
```

Authored with an AI assistant.

Co-Authored-By: Claude Fable 5.1 <noreply@anthropic.com>

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Thanks for the review. I went through each finding and pushed a new head (5fd67b045fe).

**1. "`trivial` misses the most common trailing-`.backward()` step"**
Fixed in 5fd67b045fe: the condition is now `not code.guarded_codes and not code.bypassed`, as you suggested: a non-bypassed frame with no variants ran eager during capture whether or not Dynamo entered it. `test_trivial_continuation_is_served_as_plain_python` is parametrized over `no_tensor`, `no_ops` (the `SkipFrame("No ops traced")` shape) and `closure`. I also captured your exact `return loss` step end to end at the top of the stack. There the continuation compiles a guarded code of its own, and the default gates write the artifact.

**2. "Stale comments in the driver"**
Fixed in 5fd67b045fe: the module-scope comment now reads "opened only for a frame the driver serves; a dead record never imports its module", and the `_multigraph_frames` docstring defines the trivial kind. No change needed for the "(Dynamo ran it eager as trivial)" refusal: with the widened definition, a non-entry non-bypassed frame without variants is served above it, so that branch is reached only by the entry, which did run eager.

**3. "`frame.get("trivial")` vs `frame["trivial"]`"**
Fixed in 5fd67b045fe: indexed directly in both places.

**4. "The closure branch has no test"**
Fixed in 5fd67b045fe: the `closure` case uses a module-level step that builds a lambda after `loss.backward()`, so the continuation has free variables and goes through the `_plain` factory. It is compared against eager.

**5. "The test doesn't go through `capture()`/`load()`"**
Declined here, covered above: `capture()` with the Dynamo tracer arrives in #198138. There, `test_training_step_grads_match_eager_in_a_fresh_process` captures a trailing-`.backward()` step and serves it in a fresh process, and would fail if this continuation were refused.

**Comment by bobrenjc93** [comment]
@claude review this PR. Earlier bot reviews and my replies are in this thread; please verify the fixes and dispositions.

**Comment by bobrenjc93** [comment]
Replies to the latest review. Both items are fixed.

**1. "The resume-name collision pre-check still only looks at frames with variants."**
Fixed in 8c32f90493: the pre-check now opens a module under the same condition `_make_dispatcher` opens a scope, `_frame["variants"] or (_frame["trivial"] and not _frame["is_entry"])`, so a trivial continuation recorded under an inlined function's module (`config.nested_graph_breaks`) has its resume names checked before anything is seeded. The seeding comment now says "no served frame". No dedicated test: building a trivial continuation that lands in a module no variant-bearing frame opens needs nested graph breaks across two importable modules, and the change is a condition alignment with the loop right below it.

**2. "Stale comment in `test_trivial_continuation_is_served_as_plain_python`."**
Fixed in 8c32f90493: the test now models the record `_multigraph_frames` can actually produce, a bypassed continuation (`trivial=False, bypassed=True`), and the comment says a bypassed continuation still sends the capture to installing.

@claude review this PR. Earlier bot reviews and my replies are in this thread; please verify the fixes and dispositions.

File: torch/_precompile_driver.py [hunk 1]
```diff
@@ -536,6 +536,18 @@ def _make_dispatcher(frame):
                 f"{list(target.co_freevars)!r}, which a self-contained artifact "
                 f"cannot rebuild. Regenerate it from a module-level function."
             )
+        if frame["trivial"] and not is_entry:
```
Comment (aorenste) [inline torch/_precompile_driver.py]: [blocker] My agent says:
> Gate the plain-function branch on the frame having a resume name, not on it not being the entry:
>
> ```python
>         if frame["trivial"] and frame["resume_names"]:
> ```
>
> `CompilePackage` records every user function Dynamo intercepts while the continuation runs eager, so a no-op helper called after `.backward()` becomes a trivial record with no resume names, which no frame can reach. This branch still calls `_scope` on it and imports its module at load. When the helper lives in `__main__`, the load now fails, where before this PR the capture went to installed serving:
>
> ```python
> # mylib.py
> def step(model, x, cb):
>     loss = (model(x) * 2).sum()
>     loss.backward()
>     return cb(loss)
>
> # script.py (__main__)
> def cb(t):
>     return t
> ```
>
> The records are `[step (1 variant), __resume_at_* (trivial), __main__.cb (trivial, no resume names)]`, `_serving_mode` returns `"standalone"`, and `_build_multigraph_forward()` raises "this artifact was captured from a function defined in the capturing script's __main__ module", which blames the entry even though `mylib` is importable. The entry has no resume names, so it still reaches the refusal below. Worth adding a record like `cb` to `test_trivial_continuation_is_served_as_plain_python`.

File: torch/_precompile_driver.py [hunk 2]
```diff
@@ -673,7 +685,8 @@ def dispatch(*args, **kwargs):
     opened = set()
     for _frame in frames:
         module = _frame["python_module"]
-        if _frame["variants"]:
+        # The frames _make_dispatcher opens a scope for.
+        if _frame["variants"] or (_frame["trivial"] and not _frame["is_entry"]):
```
Comment (aorenste) [inline torch/_precompile_driver.py]: [blocker] My agent says:
> Same gate here, so the pre-check opens exactly the modules `_make_dispatcher` does:
>
> ```python
>         if _frame["variants"] or (_frame["trivial"] and _frame["resume_names"]):
> ```

File: test/test_precompile.py [hunk 3]
```diff
@@ -2866,6 +2877,88 @@ def test_precompile_member_module_and_qualname_resolve_to_it(self, name):
...
+            frame["python_module"] = module
+        ns = {
+            "__name__": "precompile_test_artifact",
+            "_FRAMES": _b64(frames),
+            "_BACKENDS": _b64(backends),
+            "_ENTRY_BINDING": _b64({"defaults": None, "kwdefaults": None}),
+            "_DYNAMO_PYTHON_VERSION": tuple(sys.version_info[:2]),
+            "TORCH_VERSION": torch.__version__,
+        }
+        exec(inspect.getsource(driver._build_multigraph_forward), ns)
```
Comment (aorenste) [inline test/test_precompile.py]: [cleanup] My agent says:
> Factor the capture-then-rebuild steps into a helper shared with `test_multigraph_driver_dispatches_entry_frame` and the continuation-dispatch test after it. This is the third copy of the reset / scrub minted names / alias `sys.modules` / build `ns` / `exec` sequence. Low priority.

**Comment by bobrenjc93** [comment]
@claude review this PR. Earlier bot reviews and my replies are in this thread; please verify the fixes and dispositions.
